File: src/main/java/com/study/platform/global/exception/GlobalExceptionHandler.java
package com.study.platform.global.exception;

import com.study.platform.global.response.ApiResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.dao.PessimisticLockingFailureException;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(CustomException.class)
    public ResponseEntity<ApiResponse<Void>> handleCustomException(CustomException e) {
        log.warn("CustomException: {}", e.getMessage());
        return ApiResponse.fail(e.getErrorCode());
    }

    @ExceptionHandler(PessimisticLockingFailureException.class)
    public ResponseEntity<ApiResponse<Void>> handlePessimisticLockingFailureException(PessimisticLockingFailureException e) {
        log.warn("PessimisticLockingFailureException: {}", e.getMessage());
        return ApiResponse.fail(ErrorCode.LOCK_CONFLICT);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ApiResponse<Void>> handleValidationException(MethodArgumentNotValidException e) {
        FieldError fieldError = e.getBindingResult().getFieldErrors().get(0);
        String message = fieldError.getField() + ": " + fieldError.getDefaultMessage();
        log.warn("ValidationException: {}", message);
        return ApiResponse.fail(ErrorCode.INVALID_INPUT);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ApiResponse<Void>> handleException(Exception e) {
        log.error("UnhandledException: {}", e.getMessage(), e);
        return ApiResponse.fail(ErrorCode.INTERNAL_SERVER_ERROR);
    }
}

File: presentation/src/main/kotlin/com/study/platform/global/exception/GlobalExceptionHandler.kt
package com.study.platform.global.exception

import com.study.platform.global.response.ApiResponse
import org.slf4j.LoggerFactory
import org.springframework.dao.PessimisticLockingFailureException
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException

@RestControllerAdvice
class GlobalExceptionHandler {

    private val log = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    @ExceptionHandler(CustomException::class)
    fun handleCustomException(e: CustomException): ResponseEntity<ApiResponse<Void>> {
        log.warn("CustomException: {}", e.message)
        return ApiResponse.fail(e.errorCode)
    }

    @ExceptionHandler(PessimisticLockingFailureException::class)
    fun handlePessimisticLockingFailureException(e: PessimisticLockingFailureException): ResponseEntity<ApiResponse<Void>> {
        log.warn("PessimisticLockingFailureException: {}", e.message, e)
        return ApiResponse.fail(ErrorCode.LOCK_CONFLICT)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(e: MethodArgumentNotValidException): ResponseEntity<ApiResponse<Void>> {
        val fieldError: FieldError? = e.bindingResult.fieldErrors.firstOrNull()
        val message = if (fieldError != null) {
            "${fieldError.field}: ${fieldError.defaultMessage}"
        } else {
            e.bindingResult.allErrors.firstOrNull()?.defaultMessage ?: "유효성 검증에 실패했습니다."
        }
        log.warn("ValidationException: {}", message)
        return ApiResponse.fail(ErrorCode.INVALID_INPUT)
    }

    @ExceptionHandler(
        HttpMessageNotReadableException::class,
        MethodArgumentTypeMismatchException::class,
        MissingServletRequestParameterException::class
    )
    fun handleBadRequestException(e: Exception): ResponseEntity<ApiResponse<Void>> {
        log.warn("{}: {}", e.javaClass.simpleName, e.message)
        return ApiResponse.fail(ErrorCode.INVALID_INPUT)
    }

    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<ApiResponse<Void>> {
        log.error("UnhandledException: {}", e.message, e)
        return ApiResponse.fail(ErrorCode.INTERNAL_SERVER_ERROR)
    }
}
